### PR TITLE
major (behind the scenes) updates/hacks/workarounds to get relocatable build-cache working

### LINKED ISSUE #66: `DokkatooGenerateTask.dokkaParametersJson` contains absolute paths, making the task non-relocatable
[Build scan comparisons](https://ge.apollographql.com/c/q7k3uzw5otxyy/unxcnayzprq56/task-inputs?cacheability=cacheable,overlapping-outputs,validation-failure&expanded=WyI1Mnh0ZGNkb2s2dGhnLWRva2thbW9kdWxlZmlsZXMiLCJiY2M2NHNtYjNzYnVpLWRva2thc291cmNlc2V0cy5jb21tb25tYWluJDIuY2xhc3NwYXRoIl0#ee7filp2swu5e-dokkaparametersjson)

Example on my machine: 

```
$ cat libraries/apollo-adapters/build/dokka-config/html/dokka_parameters.json | head -n 20
{
  "moduleName": "apollo-adapters",
  "moduleVersion": "4.0.0-dev.3-SNAPSHOT",
  "offlineMode": false,
  "failOnWarning": false,
  "sourceSets": [
    {
      "sourceSetId": {
        "scopeId": ":apollo-adapters",
        "sourceSetName": "appleMain"
      },
      "displayName": "apple",
      "classpath": [
        "/Users/mbonnin/git/apollo-kotlin/libraries/apollo-api/build/classes/kotlin/iosArm64/main/klib/apollo-api.klib",
        "/Users/mbonnin/git/apollo-kotlin/libraries/apollo-annotations/build/classes/kotlin/iosArm64/main/klib/apollo-annotations.klib",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-datetime-iosarm64/0.4.0/8f2e032c52e3e59cf8e63213609007f80bdb6748/kotlinx-datetime.klib",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/com.squareup.okio/okio-iosarm64/3.2.0/a4351920c1dc049812599064a73596d245aa12c7/okio.klib",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/com.benasher44/uuid-iosarm64/0.6.0/719bcb24170ab712e6b31de02a6d6ad863bc926c/uuid.klib",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.6.21/11ef67f1900634fd951bad28c53ec957fabbe5b8/kotlin-stdlib-1.6.21.jar",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/24.0.1/13c5c75c4206580aa4d683bffee658caae6c9f43/annotations-24.0.1.jar",
```

Not really sure how to make the `/Users/mbonnin/...` relative. Maybe the project dependencies can be relative to the root project but the external ones sound tricky

### PR DESCRIPTION
**Pull request by aSemy**
- refactor Dokkatoo Configurations to share Module output directories and Module included files. This is monumentally hacky and requires several absurd workarounds, but sod it, Gradle is severely borked and they don't seem to be fixing it any time soon :(
- Deprecate DokkatooPrepareModuleDescriptorTask
- remove DokkaModuleDescriptionKxs - saving the Module descriptor to JSON broke relocatable build cache because it contained absolute file paths.
- update to Gradle 8.5
- refactor some of the internal Gradle Dependency/Task/Names helper classes to be smaller/focused/better named
- use new Gradle `Configuration.isCanBeDeclared` util, split up configurations into declarable/resolvable/consumable
- Notably I haven't added any relocatable build cache tests because I really can't be arsed

This should in theory fix #66 and #67, but this is such a massive hack I'm not sure if I should go ahead...

### Related issues

These are the Gradle issues that prevent sensible sharing of files between subprojects:

* https://github.com/gradle/gradle/issues/27594
* https://github.com/gradle/gradle/issues/27443
* https://github.com/gradle/gradle/issues/27441
* https://github.com/gradle/gradle/issues/13590
* https://github.com/gradle/gradle/issues/27265
* https://github.com/gradle/gradle/issues/23776
* https://github.com/gradle/gradle/issues/6881
* https://github.com/gradle/gradle/issues/16225
* https://github.com/gradle/gradle/issues/24131

### CODE REVIEW

File: modules/dokkatoo-plugin/src/main/kotlin/formats/DokkatooFormatTasks.kt [hunk 1]
```diff
@@ -43,63 +44,56 @@ class DokkatooFormatTasks(
...
+          this@DokkatooFormatTasks.moduleDescriptors
+        })
+      }
 
-    applyFormatSpecificConfiguration()
-  }
+      // ugly hack, workaround for https://github.com/gradle/gradle/issues/13590
+      dependsOn(providers.provider {
+        this@DokkatooFormatTasks.moduleDescriptors.map { it.moduleGenerateTaskPath }
+      })
```
Comment (aSemy) [inline modules/dokkatoo-plugin/src/main/kotlin/formats/DokkatooFormatTasks.kt:73]: And here's the end result of the ugly hack. The task paths of the module-generation tasks are extracted from the incoming Artifact Attributes and passed in as a dependency here.

**Comment by aSemy** [comment]
hey @martinbonnin, I've had a breakthrough with solving relocatable build caching of Dokka tasks - would you mind taking a look? It's such a hack I don't think I want to merge it in! But at least it's so wild it should make for good entertainment.

**Comment by martinbonnin** [comment]
Gave it a quick look, I _think_ https://github.com/adamko-dev/dokkatoo/issues/66 is not valid anymore? I couldn't find a `dokka_parameters.json` anywhere in my build using `2.0.0`. 

`module_descriptor` is still there but might be easier removing the path from there (there aren't that many)

**Comment by aSemy** [comment]
Superseded by #155

**Pull request closed without merging**